File: backend/app/domain/rules/rule_repository.py
"""Rule repository helpers ported from the frontend rules engine."""

from __future__ import annotations

import unicodedata
from typing import Any

from app.domain.rules.rule_schema import normalize_rule_atom, validate_rule_atom


class RuleRepository:
    def __init__(self, rules: list[dict] | None = None) -> None:
        self.rules: dict[str, dict] = {}
        self.by_type: dict[str, dict[str, dict]] = {}
        self.validation_errors: list[dict[str, Any]] = []
        self.add_rules(rules or [])

    @classmethod
    def from_api(cls, api: dict | None) -> "RuleRepository":
        source = (api or {}).get("source") or {}

        rules = [
            *[
                normalize_rule_atom(
                    item,
                    "feature",
                    {
                        "uuid": f"class:{_slugify(item.get('name') or 'unknown')}",
                        "tags": ["class"],
                    },
                )
                for item in ((api or {}).get("classes") or {}).values()
            ],
            *[
                normalize_rule_atom(
                    item,
                    "feature",
                    {
                        "uuid": (
                            f"feature:{_slugify(item.get('className'))}:"
                            f"{_slugify(item.get('name'))}:{item.get('level') or 1}:"
                            f"{str(item.get('source') or 'local').lower()}"
                        ),
                        "tags": ["class-feature", _slugify(item.get("className"))],
                    },
                )
                for item in source.get("classFeatures", [])
            ],
            *[
                normalize_rule_atom(
                    item,
                    "spell",
                    {
                        "uuid": (
                            f"spell:{_slugify(item.get('name'))}:"
                            f"{str(item.get('source') or 'local').lower()}"
                        ),
                        "activation": {
                            "type": _action_activation_from_casting_time(
                                item.get("castingTime")
                            ),
                            "resource_cost": (
                                "spell_slot"
                                if (item.get("level") or 0) > 0
                                else None
                            ),
                        },
                        "tags": ["spell", f"level-{item.get('level') or 0}"],
                    },
                )
                for item in (source.get("spellDetails") or {}).values()
            ],
            *[
                normalize_rule_atom(
                    item,
                    "item",
                    {
                        "uuid": (
                            f"item:{_slugify(item.get('name'))}:"
                            f"{str(item.get('source') or 'local').lower()}"
                        ),
                        "activation": {"type": "on_equip", "resource_cost": None},
                        "tags": ["item"],
                    },
                )
                for item in (source.get("itemDetails") or {}).values()
            ],
        ]

        return cls(rules)

    @property
    def size(self) -> int:
        return len(self.rules)

    def add_rules(self, rules: list[dict] | None = None) -> None:
        for rule in rules or []:
            self.add_rule(rule)

    def add_rule(self, rule: dict) -> bool:
        validation = validate_rule_atom(rule)
        if not validation["valid"]:
            self.validation_errors.append(
                {
                    "uuid": rule.get("uuid", "(unknown)") if isinstance(rule, dict) else "(unknown)",
                    "errors": validation["errors"],
                }
            )
            return False

        uuid = rule["uuid"]
        rule_type = rule["type"]
        self.rules[uuid] = rule
        self.by_type.setdefault(rule_type, {})[uuid] = rule
        return True

    def get(self, uuid: str) -> dict | None:
        return self.rules.get(uuid)

    def find_by_type(self, rule_type: str) -> list[dict]:
        return list(self.by_type.get(rule_type, {}).values())

    def find_by_tag(self, tag: str) -> list[dict]:
        return [
            rule
            for rule in self.rules.values()
            if tag in ((rule.get("metadata") or {}).get("tags") or [])
        ]


def _action_activation_from_casting_time(casting_time: Any = "") -> str:
    text = str(casting_time or "").lower()
    if "bonus" in text:
        return "bonus"
    if "reaction" in text:
        return "reaction"
    if "action" in text:
        return "action"
    return "manual"


def _slugify(value: Any) -> str:
    normalized = unicodedata.normalize("NFD", str(value or ""))
    ascii_only = "".join(char for char in normalized if not unicodedata.combining(char))

    chunks: list[str] = []
    previous_dash = False
    for char in ascii_only.lower():
        if char.isalnum():
            chunks.append(char)
            previous_dash = False
            continue
        if not previous_dash:
            chunks.append("-")
            previous_dash = True

    return "".join(chunks).strip("-")

File: src/core/rules/rule-repository.js
import { normalizeRuleAtom, validateRuleAtom } from "./rule-schema.js";

export class RuleRepository {
  constructor(rules = []) {
    this.rules = new Map();
    this.byType = new Map();
    this.validationErrors = [];
    this.addRules(rules);
  }

  static fromApi(api) {
    const source = api?.source ?? {};
    const rules = [
      ...Object.values(api?.classes ?? {}).map((item) => normalizeRuleAtom(item, "feature", {
        uuid: `class:${slugify(item.name ?? "unknown")}`,
        tags: ["class"],
      })),
      ...(source.classFeatures ?? []).map((item) => normalizeRuleAtom(item, "feature", {
        uuid: `feature:${slugify(item.className)}:${slugify(item.name)}:${item.level ?? 1}:${String(item.source ?? "local").toLowerCase()}`,
        tags: ["class-feature", slugify(item.className)],
      })),
      ...Object.values(source.spellDetails ?? {}).map((item) => normalizeRuleAtom(item, "spell", {
        uuid: `spell:${slugify(item.name)}:${String(item.source ?? "local").toLowerCase()}`,
        activation: { type: actionActivationFromCastingTime(item.castingTime), resource_cost: item.level > 0 ? "spell_slot" : null },
        tags: ["spell", `level-${item.level ?? 0}`],
      })),
      ...Object.values(source.itemDetails ?? {}).map((item) => normalizeRuleAtom(item, "item", {
        uuid: `item:${slugify(item.name)}:${String(item.source ?? "local").toLowerCase()}`,
        activation: { type: "on_equip", resource_cost: null },
        tags: ["item"],
      })),
    ];

    return new RuleRepository(rules);
  }

  addRules(rules = []) {
    rules.forEach((rule) => this.addRule(rule));
  }

  addRule(rule) {
    const validation = validateRuleAtom(rule);
    if (!validation.valid) {
      this.validationErrors.push({ uuid: rule?.uuid ?? "(unknown)", errors: validation.errors });
      return false;
    }

    this.rules.set(rule.uuid, rule);
    if (!this.byType.has(rule.type)) this.byType.set(rule.type, new Map());
    this.byType.get(rule.type).set(rule.uuid, rule);
    return true;
  }

  get(uuid) {
    return this.rules.get(uuid) ?? null;
  }

  findByType(type) {
    return [...(this.byType.get(type)?.values() ?? [])];
  }

  findByTag(tag) {
    return [...this.rules.values()].filter((rule) => rule.metadata?.tags?.includes(tag));
  }

  get size() {
    return this.rules.size;
  }
}

function actionActivationFromCastingTime(castingTime = "") {
  const text = String(castingTime).toLowerCase();
  if (text.includes("bonus")) return "bonus";
  if (text.includes("reaction")) return "reaction";
  if (text.includes("action")) return "action";
  return "manual";
}

function slugify(value) {
  return String(value ?? "")
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
}
